### ISSUE TITLE
[breaking change] discontinue .packages file

### ISSUE THREAD

**Issue by mit-mit** [dart-lang/sdk#48272]
## Intended change in behaviour

When the commands `dart pub get` or `flutter pub get` are executed, pub resolves all dependencies, and downloads those dependencies to the local pub cache. It furthermore creates a mapping from each used package to their location on the local file system. Today this mapping is stored in two files:

* `.dart_tool/package_config.json`
* `.packages` (legacy, as of Dart 2.8.0)

The former of these is the file that the current SDKs rely on. The latter has until now been generated for backwards compatibility with older Dart SDKs. After the present breaking change is made, we'll change `pub get` to only generate the former (`.dart_tool/package_config.json`).

# Rationale for making the change

The `.packages` file is left-over from older Dart SDKs and is not fully compatible with Dart 2.x SDKs (it, for example, doesn’t support [language versioning](https://dart.dev/guides/language/evolution#language-versioning)). We're removing the generation of `.packages` to prevent having extraneous files in the file system, and to prevent corner-case bugs that may occur if both files are around, but in an inconsistent state.

## Expected impact of this change.

Regular apps and packages that are written using Dart 2.x will not be impacted, as they rely on `.dart_tool/package_config.json`.

If you already have a `.packages` file, you can remote it. After upgrading to Dart 2.17 it will no longer be generated, and pub will delete it when pub get runs.

Developer tools that need to look up the packages used by a project, and their locations on the file system, will no longer be able to rely on the `.packages` file. Such tools should be upgraded to rely on the content of the `package_config.json` file. Rather than reading this file directly, we strongly recommend using the API for consuming it available in [`package:package_config`](https://pub.dev/packages/package_config).

## Steps for mitigating the change

If you for some reason depend on this legacy file, you will for 1 stable release be able to generate it by passing the flag `--legacy-packages-file` to `dart pub get` / `flutter pub get`. After one stable release, this flag will be removed, and all support for the legacy file be removed.

## Timeline

The change is planned to be made in a dev-channel release of Dart 2.17. 

The `--legacy-packages-file` flag will be removed in a dev-channel release of Dart 2.18. 

## Related issue

* https://github.com/dart-lang/pub/issues/2756

**Comment by devoncarew** [comment]
From a dart scrum discussion on this breaking change:

This change talks about not generating the `.packages` file anymore. There's another part of this however - parts of our runtime fall back on reading the .packages file if we can't find a package_config.json file. There may be parts of the ecosystem that rely on generating .packages files themselves, and then having the CFE and the analyzer consume these file.

Is removing support for reading from .packages file from our tools unspecified here? Not part of this breaking change? Something we might follow up on in a separate, future breaking change?

**Comment by devoncarew** [comment]
> This breaking change notification tracks removing the existence of `.packages`. After this breaking change lands, I'd consider any code path in our tools that falls back to `.packages` essentially dead code, and we could remove that at any time.

We should probably make that explicit in any communications to users, and, plan out when we'd remove reading the file (Dart 2.18? Dart 2.19?)

**Comment by mit-mit** [comment]
Breaking change was announced:
https://groups.google.com/a/dartlang.org/g/announce/c/jNR4yY3mcW4

**Comment by itsjustkevin** [comment]
Could you take a look at this breaking change request @grouma and @Hixie?

**Comment by grouma** [comment]
I believe this will impact a number of packages. cc @natebosch 

Off the top of my head, we rely on this file in `package:dwds` (cc @annagrin) and `package:coverage` (cc @liamappelbe).

**Comment by Hixie** [comment]
I have no objection in principle but I would not be surprised if this broke things unexpectedly (e.g. we probably have tests that depend on `.packages` in some way). Hard to determine exactly what will break without trying it though.

**Comment by natebosch** [comment]
> Off the top of my head, we rely on this file in `package:dwds`

I would have thought we'd migrated to `.dart_tool/package_config.json`, but surprisingly we haven't...

https://github.com/dart-lang/webdev/blob/9cfdbdd7c44363cf20ebb75ec3bb2ae331a7049f/dwds/lib/src/utilities/dart_uri.dart#L160

**Comment by mit-mit** [comment]
> Off the top of my head, we rely on this file in package:dwds

Filed https://github.com/dart-lang/webdev/issues/1526; @annagrin are you the right assignee for that?

> Off the top of my head, we rely on this file in ... and package:coverage 

Filed; @liamappelbe are you the right assignee for that?

**Comment by jwren** [comment]
IJ YouTrack issue is now marked as Fixed, with the PR https://github.com/JetBrains/intellij-plugins/pull/862, landed on the `master` branch.

**Comment by srawlins** [comment]
I just grepped through this repo for "dotPackages", and found references in:

* `pkg/analyzer/lib/`
* `pkg/front_end/lib/`, `pkg/front_end/test/`, and `pkg/front_end/test_cases/`
* `pkg/front_end_server/test/`
* `pkg/vm/bin/kernel_service.dart`
* `tests/lib/isolate/`, `tests/lib_2/isolate/`

**Comment by mit-mit** [comment]
@srawlins can you move that comment to https://github.com/dart-lang/sdk/issues/48275, please?

**Comment by vsmenon** [comment]
@mit-mit - are we still targeting this for the upcoming beta / stable?  Or kicking to the next one?

**Comment by mit-mit** [comment]
No, we're punting this to 2.18, and will try to make the change very early in that release cycle rather than very late in the current one.

**Comment by cbracken** [comment]
I've filed https://github.com/flutter/flutter/issues/103775 for where we're relying on .packages files in the `flutter` tool `CoverageCollector`. Discovered during a Flutter [dependencies roll](https://github.com/flutter/flutter/pull/103771).

**Comment by mit-mit** [comment]
Closing as completed.

The only remaining future work is to remove the temporary flag in pub; https://github.com/dart-lang/pub/issues/3496 tracks this.
